### ISSUE TITLE
game crash when a free attacking knight engages a walking/free walking knight and wins

### ISSUE THREAD

**Issue by sternenzaehler** [freeserf/freeserf#460]
As described, you need a free attacking knight to engage a walking on the road or a free walking enemy knight and win the battle. Immediately after that there will be a crash in viewport.cc
`Viewport::draw_serf_row_behind(MapPos pos, int y_base, int cols, int x_base) {
  for (int i = 0; i < cols;
       i++, x_base += MAP_TILE_WIDTH, pos = map->move_right(pos)) {
    /* Active serf */
    if (map->has_serf(pos)) {
      Serf *serf = interface->get_game()->get_serf_at_pos(pos);
`
this line will produce a memory access violation because the map still has a serf on and this serf has already been deleted.  So has_serf returns true and get_serf_at_pos returns nullptr.

the problem seems to be here serf.cc
`Serf::handle_knight_attacking_victory_free() {
`

when deleting the losing serf its map pos is not updated.

See for comparison serf.cc when the free attacking knight loses the battle.
`Serf::handle_serf_knight_attacking_defeat_free_state() {
...
/* Remove itself. */
    game->get_map()->set_serf_index(pos, other->index);
    game->delete_serf(this);
`

here we see that the map pos of the deleted serf is updated (though it is strange algorithm because it is not set zero but to the index of the winning serf)

adding a get_map()->set_serf_index(pos,0) in the first winning case fixes this immediate crash but there are still more bugs with free fighting that are harder to pinpoint.

If you arrange a large battle even with this fix the game will inevitable crash after some short time because the map still has stale references to the serfs that are already deleted. I will post more details here as soon as I dig deeper into this.

**Comment by sternenzaehler** [comment]
I located the bug. @wdigger this is for you) the fix wont require a pull request - just add the underscore. 

serf.cc line 4613 handle_knight_free_walking()
`            pos = map->move_left(pos_);`

here it should be "pos_ = map->move" with underscore and not "pos".

short explaination. The logic of engaging other knights requires that the field to the left of the battlefield (which is "pos_") should be passable. that is why this field is moved left to test whether it is passable.  But anyway changing the serf->pos "just so" without any bookkeeping is plain wrong and is a code smell. 
I tested it, this seems to fix the bug (tested in a battle of 600 knights!)

**Comment by sternenzaehler** [comment]
The first message in this thread is now obsolete, that code is ok, it was just causing the crash because of the "pos_" bug.

**Comment by wdigger** [comment]
@sternenzaehler do you have save before crash? Please, attach it if you have one. I prefer to check fix before commit.

**Comment by sternenzaehler** [comment]
Well, it is difficult to directly reproduce the crash because you do not directly control which knight engages first. this bug only appears if the knight that engaged wins. this may happen sooner or later or even never (if by any chance all the knights who won were defending free knights the battle may finish without a crash).
here is a save game that is almost guaranteed to reproduce the bug after maybe 2 minutes.
@wdigger 

[sm55.zip](https://github.com/freeserf/freeserf/files/3968566/sm55.zip)

**Comment by tlongstretch** [comment]
sternenzaehler, are you "playing against yourself" as multiple players, switching between them?  Who are you fighting against with no AI?

**Comment by sternenzaehler** [comment]
Yep))) thats what I do) while waiting for your AI to be ready)

**Comment by sternenzaehler** [comment]
actually what I was trying to do is to improve the engine. There are numerous bugs there that manifest themselves only when there are many knigts battling.
One of the bugs I have resolved in this issue. But there are several more
1) a dead knight crawling (because it is dead/dead animation) along the roads screaming constantly (yes I have a save game to prove this). How it happened - I do not know
2) knights hanging forever in "wait for the other to join the battle state" -they just stay on the map, the do not move and other knights ignore them. Resolved by building a road under such a knight and then removing the road. Again, I have not been able to pinpoint this bug yet.

To this purpose I played 4 players by myself on a map 3. 

I will pinpoint this bugs and document them here later.

There is one more bug I found and analysed (and fixed for myself)  -when a serf kills the serf that happens to be next in the vector:) causes nullptr exception - I messsaged wdigger here about that too.

**Comment by sternenzaehler** [comment]
also I have seen a bug that a serf can just lock itself on a road with no obstacle in front of him. this may be due to a stale index at a map pos - would be invisible in the GUI but would render the cell impassable. This happened only once or twice and is very hard to pinpoint. next time I test I would built in a cross check of all map cells versus actual serf->pos that runs regularly to see if there are discrepancies.

**Comment by tlongstretch** [comment]
I'm not surprised at all that combat bugs exist given that executing combat probably wasn't tested much with no computer AI to play against.

**Comment by sternenzaehler** [comment]
Yes that is right, it is a little bit cumbersome to test a really large combat, but I also wanted to see how it looks like:) 
I was surprised that the combat code actually works very well! It handles everything ok and the bugs I found till now are really edge cases. Ok this one pos_ bug was not really an edge case, but still not a common occurrence. So a small fight would work OK even without my proposed fixes, it is only a large scale battle that shows the bugs.

**Comment by tlongstretch** [comment]
sternenzaehler's fix definitely seems to work.  Now that I have started doing battle testing I was seeing constant issues with Free battles, now it seems to work just fine since adding that understore to 'pos'

**Comment by sternenzaehler** [comment]
> 
> 
> sternenzaehler's fix definitely seems to work. Now that I have started doing battle testing I was seeing constant issues with Free battles, now it seems to work just fine since adding that understore to 'pos'

thank you for validating this!
it seems to me we do need to do a pull request for @wdigger with this fix.

also I would say beware of this bug (described earlier in a separate thread)
quote
There is one more bug I found and analyzed (and fixed for myself) -when a serf kills the serf that happens to be next in the vector:) causes nullptr exception - I messaged wdigger here about that too.
unquote

this is not a common bug but neither is it very rare. When fights break out it is very possible that a serf kills another serf that happened to be just next in the vector (because serfs are generated sequentially for all players chances are high that serfs in the vector are interleaved between players and if you fight then the majority or all of the new serfs are knights for all players)
so if you see a null pointer exception during a battle, this is it.
I proposed a fix in a separate issue here, not sure if it was included in the main repo though.

**Comment by tlongstretch** [comment]
i'll revisit that one at some point and try your fix suggestion
